File: 06022023/01_07_Swap_Items_In_A_Dictionary/python/swap_items_in_a_dictionary.py
# https://www.codewars.com/kata/5a21e090f28b824def00013c
# Level 7 kyu

# In this kata, you will take the keys and values of a dict and swap them around.
# You will be given a dictionary, and then you will want to return a dictionary with the old values as the keys, 
# and list the old keys as values under their original keys.

# For example, given the dictionary: 
    # {'Ice': 'Cream', 'Age': '21', 'Light': 'Cream', 'Double': 'Cream'}
# Should return: 
    # {'Cream': ['Ice', 'Double', 'Light'], '21': ['Age']}

# Notes:
# The dictionary given will only contain strings
# The dictionary given will not be empty
# You do not have to sort the items in the lists

# Function
def switch_dict(before_dict: dict[str, list[str]]) -> dict[str, list[str]]:
    """Takes in a dictionary and switches the key and values for the return dictionary."""
    after_dict: dict[str, list[str]] = {dict[key]: [key] for key in dict}

    for key, value in dict.items():
        if value in after_dict.keys() and key not in after_dict[value]:
            after_dict[value].append(key)
    
    return after_dict

# Tests
# before = {
#           'Ice': 'Cream',
#           'Age': '21',
#           'Light': 'Cream',
#           'Double': 'Cream'
#           }

# expected_ans = {
#                 'Cream': ['Ice', 'Double', 'Light'],
#                 '21': ['Age']
#                 }

# usr_ans = switch_dict(before)

# # Sort lists inside dict
# usr_ans = {k: sorted(usr_ans[k]) for k in usr_ans}
# expected_ans = {k: sorted(expected_ans[k]) for k in expected_ans}

# test.describe('Basic tests')
# test.it('Sample case')
# test.assert_equals(usr_ans, expected_ans)

File: 06022023/01_07_Swap_Items_In_A_Dictionary/java/SwapItemsInADictionary.java
import java.util.*;

public class SwapItemsInADictionary {
    public static Map<String, List<String>> switchDictionary(Map<String, String> beforeDict) {
        Map<String, List<String>> afterDict = new HashMap<>();

        for (Map.Entry<String, String> entry : beforeDict.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();

            if (!afterDict.containsKey(value)) {
                afterDict.put(value, new ArrayList<>());
            }
            afterDict.get(value).add(key);
        }

        return afterDict;
    }

    public static void main(String[] args) {
        // Test the function
        Map<String, String> beforeDict = new HashMap<>();
        beforeDict.put("Ice", "Cream");
        beforeDict.put("Age", "21");
        beforeDict.put("Light", "Cream");
        beforeDict.put("Double", "Cream");

        Map<String, List<String>> afterDict = switchDictionary(beforeDict);
        System.out.println(afterDict);
    }
}
